include/bout/indexoffset.hxx

```diff
@@ -0,0 +1,50 @@
+#ifndef __INDEXOFFSET_H__
+#define __INDEXOFFSET_H__
+
+#include "bout/mesh.hxx"
+#include "bout/region.hxx"
+
+template <typename T = Ind3D>
+struct IndexOffset {
+
+  static_assert(std::is_base_of<Ind2D, T>::value || std::is_base_of<Ind3D, T>::value,
+                "IndexOffset must be templated with either Ind2D or Ind3D");
+
+  Mesh &mesh;
+  const int nx, ny, nz;
+
+  IndexOffset(Mesh &mesh)
+      : mesh(mesh), nx(mesh.LocalNx), ny(mesh.LocalNy), nz(mesh.LocalNz) {}
+
+  /// Convenience functions for converting to (x, y, z)
+  int x(T index) const { return (index.ind / nz) / ny; }
```

Comment:
`nz` is initialised by whatever mesh thinks `nz` is -- presumably this is ok when templated on `Ind3D` but not when for `Ind2D` where effective `nz` is 1?

Reply:
Right. We'll need to specialise the whole template for `Ind2D`, which will unfortunately involve some repeated code, but I'm not sure of a better way.